docs/source/getting-started/testing.rst

```diff
@@ -0,0 +1,105 @@
+    - The test also simulates the new user logging out and then logging back in. We again check the state of the ledger and see that it's the same as before.
+    - Finally we must ``close()`` the browser page, which was opened in ``newUiPage()``, to avoid runaway Puppeteer processes after the tests finish.
+
+Selecting and interacting with UI elements
+==========================================
+
+We showed how to write a simple test at a high level, but haven't shown how to make individual actions in the app using Puppeteer.
+This was hidden in the ``login()`` and ``logout()`` functions.
+Let's see how ``login()`` is implemented.
+
+.. literalinclude:: code/ui-before/index.test.tsx
+  :language: tsx
+  :start-after: // BEGIN_LOGIN_FUNCTION
+  :end-before: // END_LOGIN_FUNCTION
+
+This looks somewhat mysterious at first.
+We first get some sort of handle to the username input element, and then use it to click into it and type the party name.
+Then we directly click the login button.
+Finally, we wait until we find we've reached the menu on the main page.
+
+The strings used to find UI elements, e.g. ``'.test-select-username-field'`` and ``'.test-select-login-button'``, are `CSS Selectors <https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Selectors>`_.
+You may have seen them before if you've done CSS styling of web pages.
+In this case we use *class selectors*, which look for CSS classes we've given to elements in our React components.
+
+Here is a snippet of the ``LoginScreen`` React component, showing the ``Form`` elements that have the added classes.
+
+.. literalinclude:: code/ui-before/LoginScreen.tsx
+  :language: tsx
+  :start-after: // BEGIN_FORM
+  :end-before: // END_FORM
+
+You can see the ``className`` attributes in the ``Input`` and ``Button`` which we select in the ``login()`` function.
+Note that you can use other features of the elements to select, such as their types and attributes.
+We just show simple class selectors here.
+
+When writing CSS selectors for your tests, you will likely need to check the structure of the rendered HTML in your app by running it manually and inspecting elements using your browser's developer tools.
+For example, the image below is from inspecting the username field using Google Chrome's developer tools.
+
+   .. figure:: images/inspect-element.png
+
+There is a subtlety to explain here due to the `Semantic UI <https://semantic-ui.com/>`_ framework we use for our app.
+Semantic UI provides a convenient set of UI elements which get translated to HTML.
+In the example of the username field above, the original Semantic UI ``Input`` is translated to nested ``div`` nodes with the ``input`` inside.
```

Comment:
Add an image showing what the nested div looks like. Easier to understand.

Reply:
It is shown in the screenshot, though a little hard to read. Added a sentence pointing out it's there.